package/MDAnalysis/analysis/msd.py

```diff
@@ -0,0 +1,335 @@
+        self._position_array : :class:`np.ndarray`
+            Array of particle positions with respect to time shape = (N_frames, N_particles, 3)
+        self.N_particles : float
+            Number of particles used in the MSD calculation
+
+        """
+        #avoid memory transfer by iterating 
+        idxs = self.u.select_atoms(self.selection).ix
+        position_array_dummy = np.zeros([len(self.u.trajectory), len(self.u.atoms), 3])
+        for idx,ts in enumerate(self.u.trajectory):
+           position_array_dummy[idx,:,:] = self.u.atoms.positions
+        self._position_array = position_array_dummy[:,idxs, :]
+        self.N_particles = self._position_array.shape[1]
+
+    
+    def run(self):
+        r""" Wrapper to calculate the MSD via either the "simple" or "fft" algorithm.
+        """
+        if self.fft == True:
+            self.timeseries = self._run_fft()
+        else:
+            self.timeseries = self._run_simple()
+
+    def _run_simple(self): # naieve algorithm without FFT
+        r""" Calculates the MSD via the naieve "windowed" algorithm.
+        
+        Parameters
+        ----------
+        self._position_array : :class:`np.ndarray`
+            Array of particle positions with respect to time shape (N_frames, N_particles, 3).
+
+        Returns
+        -------
+        self.timeseries : :class:`np.ndarray`
+            The MSD as a function of lag-time.
+
+        """
+        msds_byparticle = np.zeros([self.N_frames, self.N_particles])
+        lagtimes = np.arange(1,self.N_frames)
+        msds_byparticle[0,:] = np.zeros(self.N_particles) # preset the zero lagtime so we dont have to iterate through
+        for n in range(self.N_particles):
+            for lag in lagtimes:
+                disp = self._position_array[:-lag,n,self._dim if lag else None] - self._position_array[lag:,n,self._dim]
+                sqdist = np.square(disp, dtype=np.float64).sum(axis=1, dtype=np.float64) #accumulation in anything other than f64 is innacurate
+                msds_byparticle[lag,n] = np.mean(sqdist, dtype=np.float64)
+        self.msd_per_particle = msds_byparticle
+        msds = msds_byparticle.mean(axis=1, dtype=np.float64)
+        return msds
+
+    def _run_fft(self): #with FFT, np.f64 bit prescision required.
+        r""" Calculates the MSD via the FCA fast correlation algorithm.
+        
+        Parameters
+        ----------
+        self._position_array : :class:`np.ndarray`
+            Array of particle positions with respect to time shape (N_frames, N_particles, 3).
+
+        Returns
+        -------
+        self.timeseries : :class:`np.ndarray`
+            The MSD as a function of lagtime.
+
+        """
+        msds_byparticle = np.zeros([self.N_frames, self.N_particles])
+        reshape_positions = self._position_array[:,:,self._dim].astype(np.float64)
+        for n in range(self.N_particles):
+            msds_byparticle[:,n] = tidynamics.msd(reshape_positions[:,n,:])
```

Comment:
Can you get around this loop as well? I haven't looked at `tidynamics`.

Reply:
No, only accepts 1D or 2D arrays of shape `(n_frames,dimensionality)`.